File: language-comparison/typescript/http-server.ts
/**
 * TypeScript HTTP Server 示例
 * 运行: npx tsx http-server.ts
 */

import { createServer, IncomingMessage, ServerResponse } from 'http';

// 简单的路由类型
interface Route {
    method: string;
    path: string | RegExp;
    handler: (req: IncomingMessage, res: ServerResponse, params?: Record<string, string>) => void;
}

// 用户数据
interface User {
    id: number;
    name: string;
    email: string;
}

const users: User[] = [
    { id: 1, name: "Alice", email: "alice@example.com" },
    { id: 2, name: "Bob", email: "bob@example.com" },
];

let nextId = 3;

// 辅助函数
function sendJSON(res: ServerResponse, data: unknown, statusCode = 200): void {
    res.writeHead(statusCode, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(data));
}

function parseBody(req: IncomingMessage): Promise<unknown> {
    return new Promise((resolve, reject) => {
        let body = '';
        req.on('data', chunk => body += chunk);
        req.on('end', () => {
            try {
                resolve(body ? JSON.parse(body) : {});
            } catch {
                reject(new Error('Invalid JSON'));
            }
        });
    });
}

// 路由定义
const routes: Route[] = [
    // GET /
    {
        method: 'GET',
        path: '/',
        handler: (req, res) => {
            sendJSON(res, {
                message: 'Welcome to TypeScript HTTP Server',
                endpoints: [
                    'GET /users - 获取所有用户',
                    'GET /users/:id - 获取单个用户',
                    'POST /users - 创建用户',
                    'PUT /users/:id - 更新用户',
                    'DELETE /users/:id - 删除用户',
                ]
            });
        }
    },
    // GET /users
    {
        method: 'GET',
        path: '/users',
        handler: (req, res) => {
            sendJSON(res, users);
        }
    },
    // GET /users/:id
    {
        method: 'GET',
        path: /^\/users\/(\d+)$/,
        handler: (req, res, params) => {
            const id = parseInt(params?.id || '0');
            const user = users.find(u => u.id === id);
            if (user) {
                sendJSON(res, user);
            } else {
                sendJSON(res, { error: 'User not found' }, 404);
            }
        }
    },
    // POST /users
    {
        method: 'POST',
        path: '/users',
        handler: async (req, res) => {
            try {
                const body = await parseBody(req) as Partial<User>;
                if (!body.name || !body.email) {
                    sendJSON(res, { error: 'Name and email required' }, 400);
                    return;
                }
                const newUser: User = {
                    id: nextId++,
                    name: body.name,
                    email: body.email,
                };
                users.push(newUser);
                sendJSON(res, newUser, 201);
            } catch (error) {
                sendJSON(res, { error: 'Invalid request body' }, 400);
            }
        }
    },
    // PUT /users/:id
    {
        method: 'PUT',
        path: /^\/users\/(\d+)$/,
        handler: async (req, res, params) => {
            const id = parseInt(params?.id || '0');
            const index = users.findIndex(u => u.id === id);
            if (index === -1) {
                sendJSON(res, { error: 'User not found' }, 404);
                return;
            }
            try {
                const body = await parseBody(req) as Partial<User>;
                users[index] = { ...users[index], ...body, id };
                sendJSON(res, users[index]);
            } catch {
                sendJSON(res, { error: 'Invalid request body' }, 400);
            }
        }
    },
    // DELETE /users/:id
    {
        method: 'DELETE',
        path: /^\/users\/(\d+)$/,
        handler: (req, res, params) => {
            const id = parseInt(params?.id || '0');
            const index = users.findIndex(u => u.id === id);
            if (index === -1) {
                sendJSON(res, { error: 'User not found' }, 404);
                return;
            }
            const deleted = users.splice(index, 1)[0];
            sendJSON(res, { message: 'Deleted', user: deleted });
        }
    },
];

// 路由匹配
function matchRoute(method: string, path: string): { route: Route; params: Record<string, string> } | null {
    for (const route of routes) {
        if (route.method !== method) continue;
        
        if (typeof route.path === 'string') {
            if (route.path === path) {
                return { route, params: {} };
            }
        } else {
            const match = path.match(route.path);
            if (match) {
                return { route, params: { id: match[1] } };
            }
        }
    }
    return null;
}

// 创建服务器
const server = createServer(async (req, res) => {
    const method = req.method || 'GET';
    const url = new URL(req.url || '/', `http://${req.headers.host}`);
    const path = url.pathname;
    
    // 日志
    console.log(`${new Date().toISOString()} ${method} ${path}`);
    
    // CORS 头
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    
    if (method === 'OPTIONS') {
        res.writeHead(204);
        res.end();
        return;
    }
    
    // 路由匹配
    const matched = matchRoute(method, path);
    if (matched) {
        try {
            await matched.route.handler(req, res, matched.params);
        } catch (error) {
            console.error('Handler error:', error);
            sendJSON(res, { error: 'Internal Server Error' }, 500);
        }
    } else {
        sendJSON(res, { error: 'Not Found' }, 404);
    }
});

const PORT = 3000;
server.listen(PORT, () => {
    console.log(`🚀 TypeScript HTTP Server running at http://localhost:${PORT}`);
    console.log('Available endpoints:');
    console.log('  GET    /         - API 信息');
    console.log('  GET    /users    - 获取所有用户');
    console.log('  GET    /users/:id - 获取单个用户');
    console.log('  POST   /users    - 创建用户');
    console.log('  PUT    /users/:id - 更新用户');
    console.log('  DELETE /users/:id - 删除用户');
});

File: language-comparison/go/http_server.go
// Go HTTP Server 示例
// 运行: go run http_server.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"sync"
	"time"
)

// User 用户模型
type User struct {
	ID    int    `json:"id"`
	Name  string `json:"name"`
	Email string `json:"email"`
}

// 全局数据存储
var (
	users = []User{
		{ID: 1, Name: "Alice", Email: "alice@example.com"},
		{ID: 2, Name: "Bob", Email: "bob@example.com"},
	}
	nextID = 3
	mu     sync.RWMutex
)

// 辅助函数
func sendJSON(w http.ResponseWriter, data interface{}, status int) {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(data)
}

func parseJSON(r *http.Request, v interface{}) error {
	return json.NewDecoder(r.Body).Decode(v)
}

// 日志中间件
func loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		log.Printf("%s %s %s %v", r.RemoteAddr, r.Method, r.URL.Path, time.Since(start))
	})
}

// CORS 中间件
func corsMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")

		if r.Method == "OPTIONS" {
			w.WriteHeader(http.StatusNoContent)
			return
		}

		next.ServeHTTP(w, r)
	})
}

// 路由处理器
func indexHandler(w http.ResponseWriter, r *http.Request) {
	sendJSON(w, map[string]interface{}{
		"message": "Welcome to Go HTTP Server",
		"endpoints": []string{
			"GET /users - 获取所有用户",
			"GET /users/{id} - 获取单个用户",
			"POST /users - 创建用户",
			"PUT /users/{id} - 更新用户",
			"DELETE /users/{id} - 删除用户",
		},
	}, http.StatusOK)
}

func getUsersHandler(w http.ResponseWriter, r *http.Request) {
	mu.RLock()
	defer mu.RUnlock()
	sendJSON(w, users, http.StatusOK)
}

func getUserHandler(w http.ResponseWriter, r *http.Request, id int) {
	mu.RLock()
	defer mu.RUnlock()

	for _, user := range users {
		if user.ID == id {
			sendJSON(w, user, http.StatusOK)
			return
		}
	}
	sendJSON(w, map[string]string{"error": "User not found"}, http.StatusNotFound)
}

func createUserHandler(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Name  string `json:"name"`
		Email string `json:"email"`
	}

	if err := parseJSON(r, &input); err != nil {
		sendJSON(w, map[string]string{"error": "Invalid JSON"}, http.StatusBadRequest)
		return
	}

	if input.Name == "" || input.Email == "" {
		sendJSON(w, map[string]string{"error": "Name and email required"}, http.StatusBadRequest)
		return
	}

	mu.Lock()
	newUser := User{
		ID:    nextID,
		Name:  input.Name,
		Email: input.Email,
	}
	nextID++
	users = append(users, newUser)
	mu.Unlock()

	sendJSON(w, newUser, http.StatusCreated)
}

func updateUserHandler(w http.ResponseWriter, r *http.Request, id int) {
	var input struct {
		Name  string `json:"name"`
		Email string `json:"email"`
	}

	if err := parseJSON(r, &input); err != nil {
		sendJSON(w, map[string]string{"error": "Invalid JSON"}, http.StatusBadRequest)
		return
	}

	mu.Lock()
	defer mu.Unlock()

	for i := range users {
		if users[i].ID == id {
			if input.Name != "" {
				users[i].Name = input.Name
			}
			if input.Email != "" {
				users[i].Email = input.Email
			}
			sendJSON(w, users[i], http.StatusOK)
			return
		}
	}
	sendJSON(w, map[string]string{"error": "User not found"}, http.StatusNotFound)
}

func deleteUserHandler(w http.ResponseWriter, r *http.Request, id int) {
	mu.Lock()
	defer mu.Unlock()

	for i, user := range users {
		if user.ID == id {
			users = append(users[:i], users[i+1:]...)
			sendJSON(w, map[string]interface{}{
				"message": "Deleted",
				"user":    user,
			}, http.StatusOK)
			return
		}
	}
	sendJSON(w, map[string]string{"error": "User not found"}, http.StatusNotFound)
}

// 路由器
func router() http.Handler {
	userIDPattern := regexp.MustCompile(`^/users/(\d+)$`)

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		path := r.URL.Path

		// GET /
		if path == "/" && r.Method == "GET" {
			indexHandler(w, r)
			return
		}

		// /users
		if path == "/users" {
			switch r.Method {
			case "GET":
				getUsersHandler(w, r)
			case "POST":
				createUserHandler(w, r)
			default:
				sendJSON(w, map[string]string{"error": "Method not allowed"}, http.StatusMethodNotAllowed)
			}
			return
		}

		// /users/{id}
		if matches := userIDPattern.FindStringSubmatch(path); matches != nil {
			var id int
			fmt.Sscanf(matches[1], "%d", &id)

			switch r.Method {
			case "GET":
				getUserHandler(w, r, id)
			case "PUT":
				updateUserHandler(w, r, id)
			case "DELETE":
				deleteUserHandler(w, r, id)
			default:
				sendJSON(w, map[string]string{"error": "Method not allowed"}, http.StatusMethodNotAllowed)
			}
			return
		}

		sendJSON(w, map[string]string{"error": "Not found"}, http.StatusNotFound)
	})
}

func main() {
	// 应用中间件
	handler := loggingMiddleware(corsMiddleware(router()))

	port := ":8080"
	fmt.Printf("🚀 Go HTTP Server running at http://localhost%s\n", port)
	fmt.Println("Available endpoints:")
	fmt.Println("  GET    /         - API 信息")
	fmt.Println("  GET    /users    - 获取所有用户")
	fmt.Println("  GET    /users/{id} - 获取单个用户")
	fmt.Println("  POST   /users    - 创建用户")
	fmt.Println("  PUT    /users/{id} - 更新用户")
	fmt.Println("  DELETE /users/{id} - 删除用户")

	log.Fatal(http.ListenAndServe(port, handler))
}
